### DISCUSSION TITLE
GraphQL type for Typescript record

### DISCUSSION THREAD

**Discussion by Tyrubias** [MichalLytek/type-graphql#814] [Q&A]
Supposed I have the following code:

```ts
@ObjectType()
class Widget {
    @Field()
    widgetName: string

    @Field()
    isWidgetCool: boolean
}

@ObjectType()
class Box {
    @Field(???)
    widgetsByOwner: Record<string, Widget>
}
```

How would I create a GraphQL type for `Record<string, Widget>` so that I can decorate `widgetsByOwner` with `@Field`?

**Comment by MichalLytek** [comment]
There's no support in GraphQL Spec for dynamic keys while having type constraints.

I would recommend converting `Record<string, Widget>` into a typed dictionary object:
```ts
@ObjectType()
class WidgetRecord {
  @Field()
  name: string;

  @Field()
  value: Widget;
}
```
And then use is in `Box` type:
```ts
@ObjectType()
class Box {
    @Field([WidgetRecord])
    widgetsByOwner: WidgetRecord[];
}
```

An alternative and universal approach is to use JSON scalar - https://github.com/taion/graphql-type-json.
However, then you have no ability to select fields of `Widget` in the query - it always returns full tree.

**Reply by Tyrubias** [reply]
Unfortunately, I'm adapting an existing API, so I can't really change the representation. Thanks anyways for the information! I'm not very familiar with the nuances of GraphQL
